ingestion-lambda/src/handler.ts

```diff
@@ -1,83 +1,86 @@
+import { tableName } from './database';
+import { createDbConnection } from './rds';
 import { s3Client } from './s3';
 
+type OperationOutcome = 'success' | 'failure';
+
+interface OperationResult {
+	sqsMessageId: string;
+	status: OperationOutcome;
+}
+
 export const main = async (event: SQSEvent): Promise<SQSBatchResponse> => {
 	const records = event.Records;
 
+	const sql = await createDbConnection();
+
 	console.log(`Processing ${records.length} messages`);
 
-	const eventBodies = records.map((record: SQSRecord) => ({
-		sqsMessageId: record.messageId,
-		snsMessageContent: record.body,
-		messageAttributes: record.messageAttributes,
-	}));
+	const results = await Promise.all(
+		records.map(
+			async ({
+				messageId,
+				messageAttributes,
+				body,
+			}): Promise<OperationResult> => {
+				const sqsMessageId = messageId;
 
-	const responses = await Promise.all(
-		eventBodies.map(
-			({ sqsMessageId: messageId, messageAttributes, snsMessageContent }) => {
-				const fingerpostMessageId =
-					messageAttributes['Message-Id']?.stringValue;
+				try {
+					const fingerpostMessageId =
+						messageAttributes['Message-Id']?.stringValue;
 
-				if (!fingerpostMessageId) {
-					console.warn(`Message ${messageId} is missing Message-Id attribute`);
+					if (!fingerpostMessageId) {
+						console.warn(
+							`Message ${sqsMessageId} is missing Message-Id attribute`,
+						);
+					}
+
+					// todo -- consider storing s3 object version in db
+					await s3Client.send(
+						new PutObjectCommand({
+							Bucket: BUCKET_NAME,
+							Key: `${fingerpostMessageId}.json`,
+							Body: body,
+						}),
+					);
+
+					// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- seems like postgres.js requires this format? https://github.com/porsager/postgres/issues/587#issuecomment-1563262612
+					const snsMessageContent = JSON.parse(body);
+					await sql`
+						INSERT INTO ${sql(tableName)}
+							(external_id, content)
+						VALUES
+							(${fingerpostMessageId ?? null}, ${snsMessageContent})
+						RETURNING id`.then((res) => {
+						if (res.length === 0) {
+							return 'failure';
+						}
+						return 'success';
+					});
+				} catch {
 					return Promise.resolve({
-						status: 'rejected',
-						messageId,
+						status: 'failure',
+						sqsMessageId,
 					});
 				}
 
-				const resp = s3Client.send(
-					new PutObjectCommand({
-						Bucket: BUCKET_NAME,
-						Key: `${fingerpostMessageId}.json`,
-						Body: snsMessageContent,
-					}),
-				);
-				return resp
-					.then(() => ({ status: 'resolved', messageId }))
-					.catch((reason) => {
-						console.error(`Failed to put ${messageId} to S3: ${reason}`);
-						return { status: 'rejected', messageId };
-					});
+				return { sqsMessageId, status: 'success' };
 			},
 		),
 	);
 
-	const batchItemFailures = responses
-		.filter(({ status }) => status === 'rejected')
-		.map(({ messageId }) => ({ itemIdentifier: messageId }));
+	await sql.end();
```

Comment:
I would perhaps suggest wrapping as much as possible of this function in a `try` (optional `catch`), but with this `await sql.end()` in the `finally` block to ensure its always fired)

Reply:
good call, thanks! done that